File: leetcode/arrays/01-two-sum/optimal_solution.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashMap = {}
        
        for index in range(len(nums)):
            num = nums[index]
            complement = target - num
            if num in hashMap:
                return [hashMap[num], index]
            else:
                hashMap[complement] = index

# Runtime: 86 ms, faster than 43.50% of Python3 online submissions for Two Sum.
# Memory Usage: 15.4 MB, less than 40.83% of Python3 online submissions for Two Sum.

File: leetcode/arrays/01-two-sum/optimal-solution.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        
        for(int i=0; i<nums.length; i++) {
            int complement = target - nums[i];
            
            if(map.containsKey(complement)) {
                return new int[] {map.get(complement), i};   
            }
            map.put(nums[i], i);
        }
        return null;
    }
}

// Runtime: 1 ms, faster than 99.64% of Java online submissions for Two Sum.
// Memory Usage: 39.2 MB, less than 69.83% of Java online submissions for Two Sum.